src/compiler/crystal/command/repl.cr

```diff
@@ -34,14 +35,25 @@ class Crystal::Command
     end
 
     if options.empty?
+      show_banner
       repl.run
     else
       filename = options.shift
       unless File.file?(filename)
         error "File '#{filename}' doesn't exist"
       end
 
+      show_banner
       repl.run_file(filename, options)
     end
   end
+
+  private def show_banner
+    return unless STDOUT.tty? && STDERR.tty? && !ENV.has_key?("CRYSTAL_INTERPRETER_SKIP_BANNER")
```

Comment:
Should we really skip the message when either output is not a TTY? What's the reason for that?

Reply:
This was Oleh's suggestion and to me it makes sense: if you're redirecting errors, like for logging them, you probably want just real errors. And if you're redirecting the stdout, well, the same logic applies, you probably do not want to get that printed.

Reply:
What are "real" errors? This is a message that we deem necessary to print to stderr when using the interpreter in the current, unstable state. It feels too much magic to make this message dependent on a TTY. Such a check is typically used for features that only work in a TTY (such as interactivity, ANSI colours). But this is just a message. We don't disable compiler warnings on a TTY either, even though it might not be a "real" error.
If you don't want to get that message, you can disable it via the environment variable. That is regardless of whether it's a TTY or not.

Reply:
As I see it: this message is just a "use under your own risk". If you are doing tty piping, I claim, _you know that already_. It's when you run it in normal circumstances that you need this reminder, `crystal i` o `crystal i file.cr`.

Reply:
> If you are doing tty piping, I claim, you know that already.

But why? Why does using an ubiquitous feature like piping output streams provide the user with knowledge about the experimental state the interpreter? This seems like a pretty random assumption.

Reply:
let's switch side, how would you do it?

Reply:
Remove `STDOUT.tty? && STDERR.tty?` and have the same behaviour, regardless of where output is piped.

Reply:
I really don't like that. It tampers with the user trying to chain the output of their program with something else, unless the var is set. What's the point of writing this message in a pipe? I really don't get it.

Reply:
It's the STDERR pipe. Not the output type. So it doesn't mess with the actual output of the program. Stderr is meant for error messages, status reports and other diagnostics. A welcome message for the interpreter fits into that genuinely. I see no reason why it would make a difference whether it goes to a tty or not. And no reason why the target of stdout should have an effect on what's written to stdout.
I find this behaviour weird and unexpected, and hard to understand for an observer. I would find it very confusing to see a message when I'm running a program in the console, and I want to save that message by piping stderr into a file but then the message is missing.

Reply:
> I find this behaviour weird and unexpected, and hard to understand for an observer. I would find it very confusing to see a message when I'm running a program in the console, and I want to save that message by piping stderr into a file but then the message isn't there anymore.

I'd find weird and unexpected to see such banner in the piped STDERR. These messages are IMO meant for interactive interactions, they're of no value for automated (piped) use-cases.